spec/abi.dd

```diff
@@ -697,11 +697,13 @@ $(GNAME TypeTuple):
 
 $(H2 $(LNAME2 function_calling_conventions, Function Calling Conventions))
 
-        $(P The $(D extern (C)) and $(D extern (D)) calling convention matches the C
-        calling convention
-        used by the supported C compiler on the host system.
-        Except that the extern (D) calling convention for Windows x86 is described here.
-        )
+        $(P Functions with $(D extern (C)) linkage matches the C calling
+        convention used by the supported C compiler on the host system.)
+
+        $(P Functions with $(D extern (D)) linkage matches the calling
+        convention used by $(D extern (C)) functions with the parameters
+        pushed in the reverse order, except for 32-bit x86 architecture, which
```

Comment:
Not really, parameters are just evaluated from left to right.

Reply:
Can you elaborate? Look into the godbolt link and issue attached to the PR description for context. Maybe I expressed myself wrongly.

Reply:
The real intent is left to right evaluation of parameters.

Dmd passing arguments in reverse is just an implementation detail, because Walter doesn't want to use too many scratch registers to hold evaluated parameters before calling the function.

Reply:
I don't think it is an implementation detail. If the specification is correct, then the compiler is behaving wrongly. I interpret implementation details as different implementation behaviour that doesn't affect specification conformance.

In practice, if I try to link a function generated by DMD with another binary generated by a compiler that generates a conformant convention call, bad things may happen, if the order matters, because the registers are reversed and therefore the parameters are passed wrongly.

Reply:
> GDC is spec conformant. ;-)
> 
> [godbolt.org/z/57KheoTKW](https://godbolt.org/z/57KheoTKW)

That is good to know and makes sense. So only LDC/DMD is wrong. What about the 32-bit x86 on GDC? Since the spec only mentions Windows, it is also conformant. I don't know about Windows x86 since I don't use Windows, but should we deviate from the standard calling convention there too?

Reply:
GCC only supports MinGW or Cygwin, so I only follow what is the ABI for that target.

Reply:
> GCC only supports MinGW or Cygwin, so I only follow what is the ABI for that target.

Doesn't MinGW try to mimic the same ABI as Windows? Anyway, it is not so explicit that way, so GDC seems to conform with it, only DMD and LDC do not. See https://github.com/dlang/dmd/pull/13287 where we can discuss the problem more in-depth and a possible fix for this.

Reply:
extern(C) code may be compatible, but MinGW uses Itanium ABI for C++, not MSVC.